enums.rb

```diff
@@ -0,0 +1,136 @@
+# rubocop: disable Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+# Description/Explanation of module enums
+module Enumerable
+  #:nodoc:
+  def my_each(index = 0)
+    return to_enum unless block_given?
+
+    while index < to_a.length
+      if instance_of?(Hash)
+        yield(keys[index], values[index])
+      else
+        yield(to_a[index])
+      end
+      index += 1
+    end
+    self
+  end
+
+  def my_each_with_index(index = 0)
+    return to_enum unless block_given?
+
+    while index < to_a.length
+      yield(to_a[index], index)
+      index += 1
+    end
+    self
+  end
+
+  def my_select(arr = [], index = 0)
+    return to_enum unless block_given?
+
+    my_each do |k|
+      if yield(k)
+        arr[i] = k
+        index += 1
+      end
+    end
+    arr
+  end
+
+  def check?(value, pattern, flag = nil)
+    if flag
+      return true if value.instance_of?(pattern)
+      return true if value.instance_of?(Complex) || value.instance_of?(Float) || value.instance_of?(Integer)
+    elsif value.to_s.match(pattern.to_s)
+      return true
+    end
+
+    false
+  end
+
+  def my_all?(pattern = nil, flag: false)
+    my_each do |k|
+      return false unless k
+
+      if block_given?
+        return false unless yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return false unless check?(k, pattern, flag)
+      end
+    end
+    true
+  end
+
+  def my_any?(pattern = nil, flag: false)
+    my_each do |k|
+      if block_given?
+        return true if yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return true if check?(k, pattern, flag)
+      elsif k
+        return true
+      end
+    end
+    false
+  end
+
+  def my_none?(pattern = nil, flag: false)
+    my_each do |k|
+      if block_given?
+        return false if yield(k)
+      elsif (flag = [Integer, Float, String, Numeric].include?(pattern)) || pattern.to_s.include?('?')
+        return false if check?(k, pattern, flag)
+      elsif k
+        return false
+      end
+    end
+    true
+  end
+
+  def my_count(comp = nil, count = 0)
+    my_each do |k|
+      count += 1 if block_given? && yield(k) || comp && k == comp
+    end
+
+    return to_a.length if !block_given? && comp.nil?
+
+    count
+  end
+
+  def my_inject(vos = nil, ops = nil)
+    mem = vos.public_send(ops, to_a[0]) if ops
+    arr = to_a
+    my_each_with_index do |k, i|
+      mem = to_a[0] if i.zero?
+      if ['+', '-', '/', '*'].include?(vos.to_s)
+        mem = mem.public_send(vos, arr[i + 1]) if i < arr.to_a.length - 1
+      elsif ops
+        mem = mem.public_send(ops, arr.to_a[i + 1]) if i < arr.to_a.length - 1
+
+      end
+      mem = yield(mem, k) if block_given?
+    end
+    mem = yield(mem, vos) if block_given? && vos
+    mem
+  end
```

Comment:
- [ ] raises a "LocalJumpError" when no block or argument is given
- [ ] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block.
```
  operation =  proc { |sum, n| sum + n } 
  p [1,2,3,4,5,6].inject(&operation) # 21
  p [1,2,3,4,5,6].my_inject(&operation) # 22
```

- [ ]  when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

```
  p range.my_inject { |prod, n| prod * n } 
# 6336269417023620425752626701263493509245341993533440000000000000

  p range.inject { |prod, n| prod * n } 
# 1267253883404724085150525340252698701849068398706688000000000000
```

- [ ]  when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

```
  p range.my_inject(2, :*) # 1267253883404724085150525340252698701849068398706688000000000000
  p range.inject(2, :*) # 2534507766809448170301050680505397403698136797413376000000000000
```